qos/evm/state.go

```diff
@@ -29,72 +31,186 @@ type ServiceState struct {
 		return err
 	}
 
-	if err := validateEndpointBlockNumber(endpoint, s.perceivedBlockNumber); err != nil {
+	// Ensure the service state's perceived block number is not ahead of the endpoint's block number.
+	if err := endpoint.validateBlockNumber(s.perceivedBlockNumber); err != nil {
 		return err
 	}
 
+	// TODO_IN_THIS_PR(@commoddity): #PUC
+	if s.shouldPerformArchivalCheck() {
+		if err := endpoint.validateArchivalCheck(s.archivalState.balance, endpointAddr); err != nil {
+			return err
+		}
+	}
+
+	// The service state
 	return nil
 }
 
-// UpdateFromObservations updates the service state using estimation(s) derived from the set of updated endpoints.
+// UpdateFromEndpoints updates the service state using estimation(s) derived from the set of updated endpoints.
 // This only includes the set of endpoints for which an observation was received.
 func (s *ServiceState) UpdateFromEndpoints(updatedEndpoints map[protocol.EndpointAddr]endpoint) error {
 	s.serviceStateLock.Lock()
 	defer s.serviceStateLock.Unlock()
 
+	// Initialize consensus map if it doesn't exist.
+	if s.archivalState.balanceConsensus == nil {
+		s.archivalState.balanceConsensus = make(map[string]int)
+	}
+
 	for endpointAddr, endpoint := range updatedEndpoints {
 		logger := s.logger.With(
 			"endpoint_addr", endpointAddr,
 			"perceived_block_number", s.perceivedBlockNumber,
 		)
 
-		// DO NOT use the endpoint for updating the perceived state of the EVM blockchain if the endpoint is not considered valid.
-		// e.g. an endpoint with an invalid response to `eth_chainId` will not be used to update the perceived block number.
-		if err := endpoint.Validate(s.chainID); err != nil {
+		// Validate the endpoint's chain ID; do not update the perceived block number if the chain ID is invalid.
+		if err := endpoint.validateChainID(s.chainID); err != nil {
 			logger.Info().Err(err).Msg("Skipping endpoint with invalid chain id")
 			continue
 		}
 
-		// TODO_TECHDEBT: use a more resilient method for updating block height.
-		// E.g. one endpoint returning a very large number as block height should
-		// not result in all other endpoints being marked as invalid.
-		blockNumber, err := endpoint.GetBlockNumber()
+		// Retrieve the block number from the endpoint.
+		blockNumber, err := endpoint.getBlockNumber()
 		if err != nil {
 			logger.Info().Err(err).Msg("Skipping endpoint with invalid block number")
 			continue
 		}
 
+		// Update the perceived block number.
 		s.perceivedBlockNumber = blockNumber
 
-		logger.With("endpoint_block_number", blockNumber).Info().Msg("Updating latest block height")
+		// Attempt to retrieve the archival balance from the endpoint.
+		balance, err := endpoint.getArchivalBalance()
+		if err != nil {
+			logger.Info().Err(err).Msg("Skipping endpoint without archival balance")
+			continue
+		}
+
+		// Only count non-empty balances toward consensus.
+		if s.archivalState.balance == "" && balance != "" {
```

Comment:
The logic in this method is becoming too complex to follow (granted it was already a good refactor candidate).

WDYT of moving all Archival related logic into a separate struct under `State`?
The "archival check" struct can focus solely on the `archivalState`.
This will also consolidate all the "archival" methods into a single struct.